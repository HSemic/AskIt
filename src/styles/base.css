html {
  font-size: 62.5%; /* 10px */
}

body {
  background-color: #dae0e6 !important;
}
