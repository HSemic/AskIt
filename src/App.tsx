import * as React from 'react';

import CssBaseline from '@mui/material/CssBaseline';

import { Routes, Route } from 'react-router-dom';

import Navbar from './components/organisms/Navbar';
import HomePage from './pages/HomePage';
import ProfilePage from './pages/ProfilePage';
import QuestionPage from './pages/QuestionPage';
import LoginPage from './pages/LoginPage';
import RegisterPage from './pages/RegisterPage';

import RequireAuth from './components/molecules/RequireAuth';
import VerticalSpacer from './components/atoms/VerticalSpacer';
import Footer from './components/molecules/Footer';

// import { hashAPassword } from './services/passwordHashingService';

const App = (): React.ReactElement => {
  // console.log(hashAPassword('test'));

  return (
    <>
      <CssBaseline />
      <Navbar />
      <VerticalSpacer />
      <Routes>
        <Route path="/" element={<HomePage />} />
        <Route
          path="profile"
          element={
            <RequireAuth>
              <ProfilePage />
            </RequireAuth>
          }
        />
        <Route path="/login" element={<LoginPage />} />
        <Route path="/register" element={<RegisterPage />} />
        <Route path="/question/:id" element={<QuestionPage />} />
        <Route
          path="profile/question/:id"
          element={
            <RequireAuth>
              <QuestionPage />
            </RequireAuth>
          }
        />
        <Route
          path="*"
          element={
            <main style={{ padding: '1rem' }}>
              <p>There's nothing here!</p>
            </main>
          }
        />
      </Routes>
      <VerticalSpacer />
      <Footer />
    </>
  );
};

export default App;
