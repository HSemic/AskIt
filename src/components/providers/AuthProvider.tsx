import * as React from 'react';
import { useState, useEffect } from 'react';

import { useDispatch, useSelector } from 'react-redux';

import {
  fetchUserByEmailAndValidateRequest,
  registerUserRequest,
  logout
} from '../../app/_redux/actions/userActions';

import { RootState } from '../../app/_redux/reducers/rootReducer';

interface AuthContext {
  loggedIn: boolean;
  login: (email: string, password: string) => void;
  logout: () => void;
  register: (
    firstName: string,
    lastName: string,
    email: string,
    password: string
  ) => void;
}

const authContext = React.createContext<AuthContext>({
  loggedIn: false,
  login: () => {
    return;
  },
  logout: () => {
    return;
  },
  register: () => {
    return;
  }
});

export function useAuth() {
  const dispatch = useDispatch();

  const { isUserLoggedIn } = useSelector((state: RootState) => state.user);

  const [loggedIn, setLoggedIn] = useState(false);

  useEffect(() => {
    setLoggedIn(isUserLoggedIn);
  }, [isUserLoggedIn]);

  return {
    loggedIn,
    login(email: string, password: string) {
      dispatch(fetchUserByEmailAndValidateRequest(email, password));
    },
    logout() {
      dispatch(logout());
    },
    register(
      firstName: string,
      lastName: string,
      email: string,
      password: string
    ) {
      dispatch(
        registerUserRequest({
          firstName: firstName,
          lastName: lastName,
          email: email,
          password: password
        })
      );
    }
  };
}

interface AuthProviderProps {
  children: JSX.Element;
}

export function AuthProvider({ children }: AuthProviderProps) {
  const auth = useAuth();

  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
}
