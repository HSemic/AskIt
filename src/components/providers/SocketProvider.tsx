import * as React from 'react';
import { useState, useEffect, createContext } from 'react';
import { io, Socket } from 'socket.io-client';

import { useSelector } from 'react-redux';
import { RootState } from '../../app/_redux/reducers/rootReducer';
import { apiUrl } from '../../app/api/askIt';

interface SocketContext {
  socket: Socket | null;
}

const socketContext = createContext<SocketContext>({
  socket: null
});

export function useSocket() {
  const [socket, setSocket] = useState<Socket | null>(null);

  const { isUserLoggedIn, loggedInUser } = useSelector(
    (state: RootState) => state.user
  );

  const [isUserLoggedInPrev, setIsUserLoggedInPrev] = useState(isUserLoggedIn);

  useEffect(() => {
    if (isUserLoggedInPrev === isUserLoggedIn) return;

    setIsUserLoggedInPrev(isUserLoggedIn);

    if (!isUserLoggedIn) {
      if (socket) {
        socket.close();
        setSocket(null);
      }
      return;
    }

    if (isUserLoggedIn && socket) return;

    console.log(isUserLoggedIn);

    const newSocket = io(apiUrl);

    loggedInUser && newSocket.emit('newUser', loggedInUser.id);

    setSocket(newSocket);
  }, [isUserLoggedIn]);

  return { socket };
}

interface SocketProviderProps {
  children: JSX.Element;
}

export function SocketProvider({ children }: SocketProviderProps) {
  const socket = useSocket();

  return (
    <socketContext.Provider value={socket}>{children}</socketContext.Provider>
  );
}

export default SocketProvider;
