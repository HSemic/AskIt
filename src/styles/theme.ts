import { createTheme, responsiveFontSizes } from '@mui/material/styles';

const config = {
  borderRadius: '10px'
};

// main theme
const theme = createTheme({
  typography: {
    fontFamily: 'Roboto, sans-serif', // font family
    htmlFontSize: 10, // Together with font-size setting in base.css, this sets 1rem to equal 10px
    h1: {
      fontSize: '3rem',
      fontWeight: 500
    }
  },
  components: {
    MuiCssBaseline: {
      styleOverrides: {
        body: {
          // backgroundColor: '#dae0e6', // body background color
          minHeight: '100vh',
          backgroundColor: '#FFFF20',
          backgroundImage: `url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='100%25' height='100%25' viewBox='0 0 800 400'%3E%3Cdefs%3E%3CradialGradient id='a' cx='396' cy='281' r='514' gradientUnits='userSpaceOnUse'%3E%3Cstop offset='0' stop-color='%236557DD'/%3E%3Cstop offset='1' stop-color='%23091033'/%3E%3C/radialGradient%3E%3ClinearGradient id='b' gradientUnits='userSpaceOnUse' x1='400' y1='148' x2='400' y2='333'%3E%3Cstop offset='0' stop-color='%238D83FF' stop-opacity='0'/%3E%3Cstop offset='1' stop-color='%238D83FF' stop-opacity='0.5'/%3E%3C/linearGradient%3E%3C/defs%3E%3Crect fill='url(%23a)' width='800' height='400'/%3E%3Cg fill-opacity='0.4'%3E%3Ccircle fill='url(%23b)' cx='267.5' cy='61' r='300'/%3E%3Ccircle fill='url(%23b)' cx='532.5' cy='61' r='300'/%3E%3Ccircle fill='url(%23b)' cx='400' cy='30' r='300'/%3E%3C/g%3E%3C/svg%3E")`,
          backgroundAttachment: 'fixed',
          backgroundSize: 'cover'
        }
      }
    },
    MuiPaper: {
      styleOverrides: {
        root: {
          borderRadius: config.borderRadius
        }
      }
    },
    MuiAppBar: {
      styleOverrides: {
        root: {
          borderRadius: 0
        }
      }
    },
    MuiList: {
      styleOverrides: {
        root: {
          borderRadius: config.borderRadius
        }
      }
    },
    MuiButton: {
      styleOverrides: {
        root: {
          borderRadius: config.borderRadius
        }
      }
    },
    MuiTypography: {
      styleOverrides: {
        root: {
          borderRadius: `${config.borderRadius} ${config.borderRadius} 0 0`
        }
      }
    }
  }
});

// applying responsive font size to the theme object before export
export default responsiveFontSizes(theme);
