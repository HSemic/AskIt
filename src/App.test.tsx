import * as React from 'react';

import { Provider } from 'react-redux';
import store from './app/_redux/store';

import renderer from 'react-test-renderer';

import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from '@mui/material/styles';

import theme from './styles/theme';

import { AuthProvider } from './components/providers/AuthProvider';

import App from './App';

it('expect to render App Component', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <AuthProvider>
            <BrowserRouter>
              <App />
            </BrowserRouter>
          </AuthProvider>
        </Provider>
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
