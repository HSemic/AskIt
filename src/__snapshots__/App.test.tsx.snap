// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render App Component 1`] = `
Array [
  <div
    className="MuiBox-root css-i9gxme"
  >
    <header
      className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed makeStyles-appBar-1 mui-fixed css-1wpviw0-MuiPaper-root-MuiAppBar-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
      >
        <div
          className="makeStyles-wrapper-4"
        >
          <a
            className="makeStyles-logoLink-3"
            href="/"
            onClick={[Function]}
          >
            <div
              className="MuiGrid-root MuiGrid-container css-1dfafci-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <img
                  alt="Logo"
                  src="logo2.jpg"
                />
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <div
                  className="MuiTypography-root MuiTypography-h6 makeStyles-logo-2 css-djclcc-MuiTypography-root"
                >
                  AskIt
                </div>
              </div>
            </div>
          </a>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-container css-vjrwn8-MuiGrid-root"
          >
            <a
              className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root makeStyles-button-5 css-oergs2-MuiButtonBase-root-MuiButton-root"
              href="/login"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
            >
              Log In
            </a>
            <a
              className="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root makeStyles-button-5 css-1ssim7y-MuiButtonBase-root-MuiButton-root"
              href="/register"
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
            >
              Sign Up
            </a>
          </div>
        </div>
      </div>
    </header>
  </div>,
  <div
    className="makeStyles-spacer-6"
  />,
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-s5hy0i-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-7 css-1a4ufwv-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-s5hy0i-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ryu5h6-MuiPaper-root"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 css-1b0nmq0-MuiTypography-root"
              >
                Log in to post questions
              </p>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-1in8k2h-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 css-1ube6tp-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-24gvns-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1wzwawo-MuiPaper-root"
            >
              <div
                className="MuiTypography-root MuiTypography-h6 makeStyles-listTitle-7 css-15ka3go-MuiTypography-root"
              >
                Top AskIt-ers
              </div>
              <ul
                className="MuiList-root MuiList-padding css-1fllfke-MuiList-root"
              />
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1wzwawo-MuiPaper-root"
            >
              <div
                className="MuiTypography-root MuiTypography-h6 makeStyles-listTitle-7 css-15ka3go-MuiTypography-root"
              >
                Hot questions
              </div>
              <ul
                className="MuiList-root MuiList-padding css-1fllfke-MuiList-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="makeStyles-spacer-6"
  />,
]
`;
